Extend the RichText <img> tag's attributes to support height and width

RazgrizHsu opened the pull request:
follow the original principle: the image line height must not exceed the RichText line height

Changes proposed in this pull request:
 * 扩充RichText组件，使<img>标签支持设定height及width，但维持基础原则"图像高不超过设定高行"

这是之前有提过的需求，你们看看是否可行

@cocos-creator/engine-admins

RazgrizHsu committed:
Extend the RichText <img> tag's attributes to support height and width

follow the original principle: the image line height must not exceed the RichText line height

RazgrizHsu committed:
fix unit test cases, follow original rule

knoxHuang commented on cocos2d/core/components/CCRichText.js:
```diff
@@ -505,23 +505,39 @@ var RichText = cc.Class({
+            var expectHeight = richTextElement.style.imageHeight;
+
+            //follow the original rule, the image height must less then lineHeight
+            if(expectHeight && expectHeight > 0 && expectHeight < this.lineHeight )
```
这里写 if (expectHeight > 0 && expectHeight < this.lineHeight ) 就行了

jareguo replied:
这里是不是漏了一种情况，如果 expectHeight > lineHeight, 并且 spriteHeight < lineHeight，那应该把图片放大到和 lineHeight 一致才对，照目前的代码来看，似乎不会放大

RazgrizHsu replied:
不好意思，的确是漏了这种情况，
因为在我的使用情境里，我只有考虑到原图缩小
看了一下旧的代码里，的确是会将原图放大到lineHeight的比例
下次的commit我会补上

knoxHuang commented on cocos2d/core/components/CCRichText.js:
```diff
@@ -505,23 +505,39 @@ var RichText = cc.Class({
+                spriteWidth = spriteWidth * scaleFactor;
+                spriteHeight = spriteHeight * scaleFactor;
+            }
+            else if(spriteRect.height > this.lineHeight) {
```
else if (spriteHeight  > this.lineHeight) {

knoxHuang commented on cocos2d/core/components/CCRichText.js:
```diff
@@ -505,23 +505,39 @@ var RichText = cc.Class({
+            //follow the original rule, the image height must less then lineHeight
+            if(expectHeight && expectHeight > 0 && expectHeight < this.lineHeight )
+            {
+                scaleFactor = expectHeight / spriteRect.height;
```
scaleFactor = expectHeight / spriteHeight;

knoxHuang commented on cocos2d/core/components/CCRichText.js:
```diff
@@ -505,23 +505,39 @@ var RichText = cc.Class({
+                spriteHeight = spriteHeight * scaleFactor;
+            }
+            else if(spriteRect.height > this.lineHeight) {
                 scaleFactor = this.lineHeight / spriteRect.height;
```
scaleFactor = this.lineHeight / spriteHeight;

knoxHuang commented on cocos2d/core/components/CCRichText.js:
```diff
@@ -505,23 +505,39 @@ var RichText = cc.Class({
+                spriteWidth = spriteWidth * scaleFactor;
+                spriteHeight = spriteHeight * scaleFactor;
             }
+            if(expectWidht && expectWidht > 0) spriteWidth = expectWidht;
```
if(expectWidth > 0) spriteWidth = expectWidth;

knoxHuang commented on cocos2d/core/components/CCRichText.js:
```diff
@@ -505,23 +505,39 @@ var RichText = cc.Class({
-            if(spriteRect.height > this.lineHeight) {
+            var spriteWidth = spriteRect.width;
+            var spriteHeight = spriteRect.height;
+            var expectWidht = richTextElement.style.imageWidth;
```
expectWidht  -> expectWidth

jareguo commented on cocos2d/core/label/CCHtmlTextParser.js:
```diff
@@ -123,38 +123,47 @@ cc.HtmlTextParser.prototype = {
-                var remainingArgument = tagName.substring(srcBegin+1).trim();
-                var imageSrc;
-                var endQuotIndex = -1;
+                var imageAttrReg = /(\s)*src(\s)*=|(\s)*height(\s)*=|(\s)*width(\s)*=|(\s)*click(\s)*=/;
```
把这个正则缓存一下， 不要每次调用 _attributeToObject 都重新编译

RazgrizHsu replied:
目前的写法大多是照抄下方匹配outline attribute的方式，
outline那个部份也没有做缓存，

如果要缓存的话，我是觉得拉到档案头部，也就是跟25行的
```js
:25 var eventRegx = /^(click)(\s)*=/;
```
就是变成
``` js
:25 var eventRegx = /^(click)(\s)*=/; //line:25
:26 var imageAttrReg = /(\s)*src(\s)*=|(\s)*height(\s)*=|(\s)*width(\s)*=|(\s)*click(\s)*=/; //line:26
```
放在一起，这样如何？

jareguo replied:
可以~

jareguo commented on cocos2d/core/label/CCHtmlTextParser.js:
```diff
@@ -123,38 +123,47 @@ cc.HtmlTextParser.prototype = {
+                    } else if (tagName === "click") {
+                        obj.event = this._processEventHandler(tagName + "=" + tagValue);
                     }
+                    header = attribute.match(imageAttrReg);
```
正则使用全局模式是否就能不用重复匹配？

RazgrizHsu replied:
这边看来是不能省略的，省略了的话，它就取不到下一个attribute的值，会变成死循环

如上面那则回覆，
因为这边使用regular expression来匹配每一个attriubte的方式
我是照抄下方匹配outline的方式，
之所以会照抄，是因为我以为这种写法是你们认可的，所以才这么写

否则的话，我应该会使用
1. 直接用regular的group来匹配每一个attribute
2. 或者第一次使用regex匹配，接着每个子attribute仅使用string index+substring来匹配
    ( 由于我不晓得在现在的js引擎里，是否substring效率高于regex，所以这两种写法会考虑下 )

jareguo replied:
之前子龙的实现其实我没怎么细看，所以会现在才提出这个疑问。那就先按你现在这样的实现吧，之后有需要再优化。

RazgrizHsu replied:
好的，謝謝 :)

RazgrizHsu committed:
according to the discussion

- correct wrong word of Width
- include case when image height less then lineHeight
- promote image attribute regex to global cache in file

jareguo commented on cocos2d/core/components/CCRichText.js:
```diff
@@ -505,23 +505,38 @@ var RichText = cc.Class({
+                scaleFactor = expectHeight / spriteHeight;
+                spriteWidth = spriteWidth * scaleFactor;
+                spriteHeight = spriteHeight * scaleFactor;
+            } else {
```
这样一来是不是会导致未定义 imageHeight 时，图片也会被拉伸到行高？

RazgrizHsu replied:
是的，我以为你要的效果是跟随原本的方式，
最初的版本里，任何图片都会缩放至与行高相同，

现在是变成，
1. 仅在你有定义image height，并且值小于行高，才替你缩小
2. 其余的情况，与最初的版本相同，会缩放到与行高相同

当然，我是觉得如果该图片原本的大小，就小于行高的话，应该是维持原本的大小比较合理，

看您怎么决定，我再修改

RazgrizHsu replied:
另外，
其实我原本是打算连：当图片设定的值大于line height时，就调整该行行高
不过因为这样改动有些大，怕你们不能接受，所以就先跟随原本的基础原则

jareguo replied:
> 最初的版本里，任何图片都会缩放至与行高相同，
> 当然，我是觉得如果该图片原本的大小，就小于行高的话，应该是维持原本的大小比较合理，

之前的实现我也不太清楚， @knoxHuang 怎么看？要不要保持和原有一致？

knoxHuang replied:
我觉得这样就可以了，有设置 image height 才会变动，其余的跟原来的版本一样就行

jareguo replied:
OK，那就这样
